src/config/mod.rs

```diff
@@ -32,13 +32,18 @@ pub const ENV_VAR_PREFIX: &str = "CT_";
 
 // Environment variable name used to specify the CloudTruth API value, so it does not need to be
 // specified on the command line.
-#[allow(dead_code)]
 pub const CT_API_KEY: &str = "CT_API_KEY";
 
 // Environment variable name used to override the default server URL.
-#[allow(dead_code)]
 pub const CT_SERVER_URL: &str = "CT_SERVER_URL";
 
+// Environment variable name used to set the environment name.
+pub const CT_ENV: &str = "CT_ENV";
+
+// List of variables to remove to make a clean environment.
+#[allow(dead_code)]
```

Comment:
While I may have originally pushed for the notion of configuration by environment (in addition to command line and config file), because that's what many CLIs do, in this case I think that may have been a mistake.

Doppler explicitly does not use environment variables to control their CLI, likely to avoid the issues we've been trying to solve here.  I believe it would be cleaner overall if we did not allow for any environment variable based configuration of our CLI, since we have a product that mucks with environment variables, and require only (command-line arguments | configuration file).  Inevitably we're going to end up with a customer who has a slew of CT_ prefixed variables that have nothing to do with us, AND they will use CT_ENV and CT_PROJECT (or worse, CT_API_KEY) which will mean nothing useful to us, and we'll eventually have wished we did not allow the CLI to be controlled by environment variables at all. :)

In other words, if you could not use CT_ anything to control our CLI, --permissive would not be necessary.  The none/overlay/underlay/exclusive control we put in earlier is still valid, however.

Reply:
In this case of a GitLabs CI/CD pipeline I think it using a sourced environment key stored as a variable.  That is how I have my test GitLab pipeline setup.  It is specifically using ``CT_API_KEY`` with the API token.  While the name collision is a concern I believe we do want to keep them for integration potentials.  We could pass in the key to the cmd line with -k in the pipeline but wouldn't you prefer the global?  I agree with Kevin on blocking the ``CT_`` prefix and expanding if required later.

Reply:
If the `CT_` prefix is problematic, we could use something less likely to conflict. `CT_CLI_` or `CLOUDTRUTH_` or `CLOUDTRUTH_CLI_`. The reason we used the short name was to operate similarly to the AWS suite of tools. Both the CLI and the client libraries will source the same set of variables. But, we don't have any client binding libraries, so using a more specific name would work. We'd need to add a warning or some such so existing users know to cut over.

The reason for configuration via env is mostly what @dio-ct indicated. It's also the dominant (or only) way of configuring tools run in Docker, Kubernetes, ECS, etc. We could force people to consume the values via CLI args (e.g., `--api-key ${SECRET_API_KEY}`), but I'm not sure that materially changes things.